src/download_px4_logfile/mavctrl/ftp.go

```diff
@@ -0,0 +1,549 @@
+	ftpNakNoSessionAvail uint8 = 5
+	ftpNakEOF            uint8 = 6
+	ftpNakUnknownCmd     uint8 = 7
+	ftpNakFileExists     uint8 = 8
+	ftpNakFileProtected  uint8 = 9
+	ftpNakFileNotFound   uint8 = 10
+)
+
+const ftpMaxPayloadLen uint32 = uint32(ftpMsgLen - ftpHdrPosData)
+
+type ftpheader struct {
+	opcode uint8
+	size   uint8
+	offset uint32
+}
+
+type session struct {
+	id     uint8
+	active bool
+}
+
+type Diritem struct {
+	Type string
+	Name string
+	Size int
+}
+
+var (
+	ftp_session = session{0, false}
+	ftp_seq     = uint16(0)
+)
+
+/**********************************************************
+ * Mavlink Control API functions
+ */
+
+func (ctrl *MavlinkControl) CheckSdCardExists() bool {
+	return ctrl.ftpCheckExists("/fs", "D", "microsd")
+}
+
+func (ctrl *MavlinkControl) CheckLogDirExists() bool {
+	if !ctrl.ftpCheckExists("/fs", "D", "microsd") {
+		log.Println("[ftp] No SDCARD found")
+		return false
+	}
+	if !ctrl.ftpCheckExists("/fs/microsd", "D", "log") {
+		log.Println("[ftp] No log dir found")
+		return false
+	}
+	return true
+}
+
+func (ctrl *MavlinkControl) ListDirectory(dir string) (list []Diritem, list_ok bool) {
+	list_ok = true
+	msg, ok := ctrl.ftpSendMsg(
+		ftpheader{
+			opcode: ftpOpcodeListDirectory,
+			size:   uint8(len(dir) + 1),
+			offset: 0,
+		}, []uint8(dir), false)
+
+	if ok {
+		data, size := ctrl.ftpGetData(msg)
+		parse_state := 0 // Wait for type
+
+		var item Diritem
+		var item_data_buf string = ""
+		for i := uint8(0); i < size; i++ {
+			switch parse_state {
+			case 0: // GET TYPE
+				if data[i] == 0x00 {
+					// null char -> start over
+					continue
+				} else {
+					// item type received --> read item name
+					item.Type = string(data[i])
+					parse_state = 1
+				}
+			case 1: // PARSE NAME
+				if data[i] == 0x00 || data[i] == 0x09 {
+					// Termination char -> add name
+					item.Name = item_data_buf
+					item_data_buf = ""
+				}
+
+				switch data[i] {
+				case 0x00:
+					// Item ready, add to list
+					list = append(list, item)
+					// null char -> start over
+					parse_state = 0
+				case 0x09:
+					// TAB char -> name ends, get file size
+					parse_state = 2
+				default:
+					// fill the name buffer
+					item_data_buf += string(data[i])
+				}
+			case 2: // GET FILE SIZE
+				if data[i] == 0x00 {
+					// Termination char -> add size
+					i, err := strconv.Atoi(item_data_buf)
+					if err != nil {
+						item.Size = 0
+						log.Printf("[ftp] ERROR to read list file size: %s\n", err.Error())
+					} else {
+						item.Size = i
+					}
+					item_data_buf = ""
+					// Item ready, add to list
+					list = append(list, item)
+					parse_state = 0
+				} else {
+					// fill the size buffer
+					item_data_buf += string(data[i])
+				}
+			}
+		}
+	}
+	return
+}
+
+func (ctrl *MavlinkControl) DownloadLogFile(src_dir string, src_file string, dst_path string, fsize int) bool {
+	// Download file
+	var src_file_path string = "/fs/microsd/log/" + src_dir + "/" + src_file
```

Comment:
Mainly cosmetic tip: there's a nice stdlib function `filepath.Join()` which can be used to build paths: https://pkg.go.dev/path/filepath#Join